src/Microsoft.DotNet.Interactive.CSharpProject/RoslynWorkspaceUtilities/RoslynWorkspaceUtilities.cs

```diff
@@ -0,0 +1,345 @@
+    {
+        // Log error
+    }
+
+    private static void WorkspaceChangedHandler(object sender, WorkspaceChangeEventArgs e)
+    {
+        // Log
+    }
+
+    private static ProjectInfo GetProjectInfo(BuildProjectData buildProjectData, CodeAnalysis.Workspace workspace, ProjectId projectId)
+    {
+        string projectName = Path.GetFileNameWithoutExtension(buildProjectData.ProjectFilePath);
+        if (!TryGetSupportedLanguageName(buildProjectData.ProjectFilePath, out string languageName))
+        {
+            return null;
+        }
+        return ProjectInfo.Create(
+            projectId,
+            VersionStamp.Create(),
+            projectName,
+            projectName,
+            languageName,
+            filePath: buildProjectData.ProjectFilePath,
+            outputFilePath: buildProjectData.TargetPath,
+            documents: GetDocuments(buildProjectData, projectId),
+            projectReferences: GetExistingProjectReferences(buildProjectData, workspace),
+            metadataReferences: GetMetadataReferences(buildProjectData),
+            analyzerReferences: GetAnalyzerReferences(buildProjectData, workspace),
+            parseOptions: CreateParseOptions(buildProjectData, languageName),
+            compilationOptions: CreateCompilationOptions(buildProjectData, languageName));
+    }
+
+    private static bool TryGetSupportedLanguageName(string projectPath, out string languageName)
+    {
+        switch (Path.GetExtension(projectPath))
+        {
+            case ".csproj":
+                languageName = LanguageNames.CSharp;
+                return true;
+            default:
+                languageName = null;
+                return false;
+        }
+    }
+
+    private static IReadOnlyList<DocumentInfo> GetDocuments(BuildProjectData buildProjectData, ProjectId projectId) =>
+        (buildProjectData
+            .SourceFiles?
+                .Where(File.Exists)
+                .Select(x => DocumentInfo.Create(
+                    DocumentId.CreateNewId(projectId),
+                    Path.GetFileName(x),
+                    loader: TextLoader.From(
+                        TextAndVersion.Create(
+                        SourceText.From(File.ReadAllText(x), Encoding.Unicode), VersionStamp.Create())),
+                    filePath: x))
+            ?? Array.Empty<DocumentInfo>()).ToList();
+
+    private static IReadOnlyList<ProjectReference> GetExistingProjectReferences(BuildProjectData buildProjectData, CodeAnalysis.Workspace workspace) =>
+        (buildProjectData.ProjectReferences
+            .Select(x => workspace.CurrentSolution.Projects.FirstOrDefault(y => y.FilePath.Equals(x, StringComparison.OrdinalIgnoreCase)))
+            .Where(x => x != null)
+            .Select(x => new ProjectReference(x.Id))
+        ?? Array.Empty<ProjectReference>()).ToList();
+
+    private static IEnumerable<MetadataReference> GetMetadataReferences(BuildProjectData buildProjectData) =>
```

Comment:
It's more appropriate here to return `IReadOnlyList<MetadataReference>`, for the same reasons as the others. Otherwise, repeat evaluations will redo expensive work.

Reply:
changed